File: 2021/star-13.py
import sys
import pprint

def fuel_required(pos, crabs):
    sum = 0
    for crab in crabs:
        sum = sum + abs(crab - pos)

    return sum

def run(filename, verbose):
    lines = []

    with open(filename) as file:
        lines = [line.strip() for line in file.readlines()]

    positions = [int(v) for v in lines[0].split(',')]
    if (verbose):
        pprint.pprint(positions)

    max = positions[0]
    min = positions[0]

    for p in positions:
        max = p if p > max else max
        min = p if p < min else min

    best = fuel_required(min, positions)
    for i in xrange(min, max + 1):
        fuel = fuel_required(i, positions)
        if (verbose):
            print('Fuel required for pos %d: %d' % (i, fuel))

        if (fuel < best):
            best = fuel
            if (verbose):
                print('Best so far is position %d with a cost of %d' % (i, fuel))

    return best

expected = 37;
actual = run('star-13-example.txt', True);
if actual == expected:
    print("Example answer OK: %d." % (actual));
else:
    print('You broke the example, expected: %d, actual: %d.' % (expected, actual))
    sys.exit()


output = run('star-13-input.txt', False);

print("The puzzle answer is: %d" % (output));

File: 2021/star-13.php
<?php

function fuel_required($pos, $crabs)
{
    $sum = 0;
    foreach ($crabs as $crab) {
        $sum += abs($crab - $pos);
    }

    return $sum;
}

function run($filename, $verbose = false)
{
    $lines = array_map('trim', file($filename));

    $positions = explode(',', array_shift($lines));
    $max = $positions[0];
    $min = $positions[0];

    foreach ($positions as $pos) {
        $max = $pos > $max ? $pos : $max;
        $min = $pos < $min ? $pos : $min;
    }

    $best = fuel_required($min, $positions);

    for ($i = $min; $i <= $max; $i++) {
        $fuel = fuel_required($i, $positions);
        if ($verbose) {
            echo "Fuel required for pos $i: $fuel\n";
        }

        if ($fuel < $best) {
            $best = $fuel;
            if ($verbose) {
                echo "Best so far is position $i with a cost of $fuel\n";
            }
        }
    }

    return $best;
}

$expected = 37;
$actual = run('star-13-example.txt', true);
if ($actual !== $expected) {
    echo 'You broke the example, ' . "expected: $expected, actual: $actual.\n";
    die();
} else {
    echo "Example answer OK: $actual\n";
}

$output = run('star-13-input.txt');

echo "The puzzle answer is:  $output\n";
